calicoctl/resourcemgr/resourcemgr.go

```diff
@@ -56,78 +83,238 @@ type ResourceActionCommand func(*client.Client, unversioned.Resource) (unversion
-	return fmt.Sprintf("Resource(%s %s)", r.typeMetadata.Kind, r.typeMetadata.APIVersion)
+	if !r.isList {
+		return fmt.Sprintf("Resource(%s %s)", r.resource.GetObjectKind(), r.resource.GetObjectKind().GroupVersionKind())
+
+	}
+	return fmt.Sprintf("Resource(%s %s)", r.listResource.GetObjectKind(), r.listResource.GetListMeta().GetResourceVersion())
 }
 
-// Store a resourceHelper for each resource unversioned.TypeMetadata.
-var helpers map[unversioned.TypeMetadata]resourceHelper
+// Store a resourceHelper for each resource.
+var helpers map[schema.GroupVersionKind]resourceHelper
+var kindToRes = make(map[string]runtime.Object)
 
-func registerResource(res unversioned.Resource, resList unversioned.Resource,
+func registerResource(res runtime.Object, resList runtime.Object, isNamespaced bool, names []string,
 	tableHeadings []string, tableHeadingsWide []string, headingsMap map[string]string,
-	apply, create, update, delete, list ResourceActionCommand) {
+	create, update, delete, get, list ResourceActionCommand) {
 
 	if helpers == nil {
-		helpers = make(map[unversioned.TypeMetadata]resourceHelper)
+		helpers = make(map[schema.GroupVersionKind]resourceHelper)
 	}
 
-	tmd := res.GetTypeMetadata()
 	rh := resourceHelper{
-		typeMetadata:      tmd,
+		resource:          res,
 		resourceType:      reflect.ValueOf(res).Elem().Type(),
 		tableHeadings:     tableHeadings,
 		tableHeadingsWide: tableHeadingsWide,
 		headingsMap:       headingsMap,
 		isList:            false,
-		apply:             apply,
+		isNamespaced:      isNamespaced,
 		create:            create,
 		update:            update,
 		delete:            delete,
+		get:               get,
 		list:              list,
 	}
-	helpers[tmd] = rh
+	helpers[res.GetObjectKind().GroupVersionKind()] = rh
 
-	tmd = resList.GetTypeMetadata()
 	rh = resourceHelper{
-		typeMetadata:      tmd,
+		listResource:      resList.(ResourceListObject),
 		resourceType:      reflect.ValueOf(resList).Elem().Type(),
 		tableHeadings:     tableHeadings,
 		tableHeadingsWide: tableHeadingsWide,
 		headingsMap:       headingsMap,
 		isList:            true,
 	}
-	helpers[tmd] = rh
+	helpers[resList.GetObjectKind().GroupVersionKind()] = rh
+
+	for _, v := range names {
+		kindToRes[v] = res
+	}
+}
+
+func (rh resourceHelper) GetObjectType() reflect.Type {
+	return rh.resourceType
+}
+
+// Apply is an un-typed method to apply (create or update) a resource. This calls Create
+// and if the resource already exists then we call the Update method.
+func (rh resourceHelper) Apply(client client.Interface, resource runtime.Object) (runtime.Object, error) {
+
+	// Store the original ResourceVersion for the Update operation later.
+	originalRV := ""
+
+	// Try to assert the resource type and remove the resourceVersion, because
+	// Create call can't have resourceVersion set and Update automatically
+	// gets and sets the resourceVersion to the latest one from the datastore.
+	_, ok := resource.(ResourceObject)
```

Comment:
This logic could be better expressed:

```
if res, ok := resource.(ResourceObject); ok {
  res.GetObjectMeta().SetResourceVersion("")
}
```

so that you don't have to cast multiple times.

Reply:
However...

The Apply/Update etc. methods deal with ResourceObjects and not ResourceListObjects therefore this check should be redundant.  Instead, it would make more sense to change the signature of this method to make that clearer - so use ResourceObject rather than runtime.Object.

Reply:
With a kubectl Apply - does it ignore the RV if the resource doesn't exists (i.e. it's doing a create) and use the RV is it does exists (and is therefore doing an update), or does it basically assume that if an RV is specified it will always be an Update?